### ISSUE TITLE
Collaboration with OKLetsPlay: LetsBet integration

### ISSUE THREAD

**Issue by qreeves** [redeclipse/world#79]
As I've been discussing with the active community on [Discord](https://redeclipse.net/discord) (join it if you haven't already!), I've been approached by a company called [OKLetsPlay](https://okletsplay.com/) who are looking for games to incorporate their LetsBet integration, which in basic terms allows players to arrange matches with others and bet on the outcome.

What's interesting about LetsBet is how it is using skill based betting (which isn't gambling because chance is not part of the outcome) without trying to shove it down the throats of the people who play the game. OKLP has been especially great about answering my questions, and providing all the information I can share with you guys. They are also aware that transparency is very important to an open source community like ours.

Some excerpts from their introduction email:

>The integration itself is entirely on the back-end of the game: we let the Red Eclipse back-end know when there's a match between two players and you let us know when the match is being played and who won. We don't operate within the UI of the game and players who aren't interested in betting don't even need to know it's supported.
>
>We will be hosting tournaments, integrating with Twitch, Hitbox, and other streams, and use cross-game promotion to bring even more users into Red Eclipse.
>
>We have no interest in buying out Red Eclipse or taking anything over. Red Eclipse will remain free, open source, and under your control. Our integration is very light! However, we can bring in additional users, keep them engaged, and provide you with a significant increase in monthly revenue.

They've also provided me with a developer presentation which can be shared with the public, should anyone be interested in their overall plan and reasoning: [LetsBet Presentation Deck Oct 15.pdf](https://github.com/red-eclipse/world/files/2481016/LetsBet.Presentation.Deck.Oct.15.pdf)

I know what you're probably thinking, "Has Quin sold out?", and I have to answer that with a profound "NO". People who are familiar with Red Eclipse know that it has been my project for over ten years now, and I've never really been motivated by monetary gain. I'm happy with our game and our community, and I want to assure you that this will not impact negatively on the game you know and love. You won't be spammed, and you won't be lured into spending money when playing the game normally.

Money, hot topic, I know. I can disclose that I will be getting paid for my time writing the integration, and if everything works out there will be an ongoing revenue stream. From the feedback I've gotten so far, people seem happy for me to see some gain from this, considering my long term dedication in keeping the project fully free and open source, but it is something I want to put out there and make sure everyone is aware of.

So what's involved? There's some prep work I'll have to do before starting on the integration which will mostly consist of:
- Improving the server's ability to referee matches in order to mitigate cheating (biggest job).
- Implementing an SSO/OpenID endpoint for Red Eclipse accounts.
- Replacing randomness with more skill based elements (which I was already doing anyway).
- Preparing the master server for things like intercommunication, matchmaking, etc.

Feel free to share your thoughts, questions, concerns, and/or suggestions. If I can't provide an answer for something I will try to get one for you if appropriate. Heck, I might even be able to rent proper servers finally.

**Comment by MirceaKitsune** [comment]
Interesting and worth following on. As long as this doesn't risk dragging the project into the crosshairs of governments and what not, and obviously it doesn't affect the average player, it sounds like a cool idea! I've actually been thinking about such a system in games, though my main thought was builtin cryptocurrency betting.

**Comment by nok-ko** [comment]
Super cool, RE is moving up in the world!

**Comment by lambda-11235** [comment]
I don't know if this was mentioned, but one source of randomness of particular concern is bots. I know that having the bots that act differently can really affect a game. Some get stuck in a corner while others protect a base for example, which gives a disadvantage to one team.

**Comment by jeaye** [comment]
> I don't know if this was mentioned, but one source of randomness of particular concern is bots. I know that having the bots that act differently can really affect a game. Some get stuck in a corner while others protect a base for example, which gives a disadvantage to one team.

That's a good thing to note. However, LetsBet matches will only be applying to players' own skill, not the skill of bots. So, to start with, this means only 1:1 matches. Onward, considering team games, this would mean that one would be on one's own score, not the score of the team. So, for example, a team-based ictf match would support players betting on their own capture counts, frag counts, etc. This helps rule out the issue of teammates throwing games (which most bots can be considered doing anyway).

We are interested in adding side betting support (using a closed-loop currency, to ensure legality). With that, both spectators and players could bet on the game outcomes, including the performance of the whole team. Legally, this side betting is gambling, as opposed to the betting on one's own skill; that's why it needs to be done with a closed-loop currency which one cannot cash out for money. Instead, those credits could be used in-app or in participating services.

**Comment by qreeves** [comment]
Yeah, we covered things like this in our first meeting, but I haven't really made public mention of any of the things I plan to do to address these issues once I start working on the integration. The answer from @jeaye is probably better than the one I would've given (thanks for fielding that one 😄).

I've already audited the use of randomness in the code, and plan to reduce how much of a role it plays in the game entirely (so skill plays a better role in the outcome of a match), but there will also be an extra level of restrictions placed on LetsBet matches (like no bots, etc).

**Comment by ds6** [comment]
I played Sauerbraten back in the day, and it's amazing to see you still working on it @qreeves. You definitely deserve a revenue stream for all this, and having one while keeping the game free is fantastic. I imagine interested parties will love being able to make money for playing well. It sounds like a great idea.
As for the side betting, I could imagine the addition of the usual microtransaction-y features like skins purchased with in-game currency to help attract a lot of new players that expect that kind of functionality with today's games. Having it be based on in-game betting would be pretty cool and a deviation from the usual models.
For a bit of wishful thinking, it would be cool to also support whole-team bets, where a player can encourage their team to do better by also including them in the pot without having them actually contribute... Maybe something like split up 20% of your bet so that you get 60% of earnings back from that 20% and the last 40% is divvied up between your teammates based on how much they contributed to the team.

**Comment by noahb01** [comment]
I'm gonna have to give a hard no on this one. This is an open source project and I feel like adding literal micro-transactions to it both detracts from the open-source nature of this project and opens up a very slippery slope towards more pay-to-win style features as well as closed source elements. How do you keep clients from faking the skins they have? How do you keep server owners from simply disabling them? Does LetsBet have an open source lib or will we need to link to a proprietary blob? Will RE have to follow a LetsBet TOS for using their service? Will players need to accept this TOS to play?

Also, RE is already pretty tight on developers. There just simply isn't the manpower to be making skins. RE would just turn into Fortnite: stagnation of the core game and an absurd number of skins, b/c thats where the money is.

If this is added, I can easily see a fork and a major split in the community, which will be damaging to both sides. Personally, provided there is no way to fully and completely disable any and all micro-transactions and money-based features on my server, I will shut it down immediately. I will not support this project if micro-transactions is dragged into it, and I suspect many others would do the same.

Oh, and then there's the legal issues. What happens when someone sues RE because they believe they lost money on something they shouldn't have? Somehow I don't get the feeling that RE could finance a proper legal defense.

**Comment by qreeves** [comment]
@noahb01 Maybe read the post properly before commenting.

**Comment by noahb01** [comment]
@qreeves I'm replying to the previous comment, specifically the part about adding micro-transactions.
Sorry if that's not clear, I'll edit my comment.

**Comment by qreeves** [comment]
@noahb01 Ah OK. Red Eclipse will never have microtransactions anyway, and talking about it in this thread is fairly off-topic.

**Comment by jeaye** [comment]
As for OkLetsPlay, which is on topic here, we're an external app players can opt into installing. The game server doesn't link with us, it just uses our HTTP API and player's servers wouldn't use it, since we don't trust them; only "official" OLP RE servers would (which is something we'll happily work out, including helping host those, as needed).

Legally, OLP assumes all liability of actions taken while using our app with any of our partner games.

Some updated info on OLP:

* Our site: https://okletsplay.com
* Our Discord: https://discordapp.com/invite/GvpZhNq
* Our latest Rocket League streamer montage: https://www.youtube.com/watch?v=Xul54Uhuuco

We're going live with our real money hard launch for Rocket League tomorrow and we're still very much interested in bringing RE onboard.

**Comment by ds6** [comment]
> I'm gonna have to give a hard no on the micro-transactions. This is an open source project and I feel like adding micro-transactions to it both detracts from the open-source nature of this project and opens up a very slippery slope towards more pay-to-win style features as well as closed source elements.

Please re-read was written in the OP and then what I said:
>We are interested in adding side betting support (using a closed-loop currency, to ensure legality). With that, both spectators and players could bet on the game outcomes, including the performance of the whole team. Legally, this side betting is gambling, as opposed to the betting on one's own skill; that's why it needs to be done with a closed-loop currency which one cannot cash out for money. Instead, those credits could be used in-app or in participating services.

> As for the side betting, I could imagine the addition of the usual microtransaction-y features like skins purchased with in-game currency

In no way was I vying for outright paid microtransactions outside of the 1:1 betting. That's awful.
I think OLP is a good idea for generating revenue without compromising the free software spirit. But if side bets were implemented the resulting currency would have to be used on *something*, and I think it would help attract players to include cosmetic items (re: not affecting gameplay).
And about preventing item spoofing, it's my belief that server owners should be able to do what they want, including faking items. On compliant/official servers item inventories should be tracked on a central server so that clients cannot spoof.
But I'm just spitballing, and @.qreeves is right that this is mostly off topic to OLP.
